File: 03_liskov_substitution/dart/bad/index01.bad.dart
class Rectangle {
  Rectangle(this.width, this.height);

  late int height;
  late int width;

  void setWidth(int width) => this.width = width;

  void setHeight(int height) => this.height = height;

  int area() => this.width * this.height;
}

class Square extends Rectangle {
  Square(super.width, super.height);

  setWidth(int width) {
    this.width = width;
    this.height = width;
  }

  setHeight(int height) {
    this.width = height;
    this.height = height;
  }
}

void increaseReactangleWidth(Rectangle rectangle) {
  rectangle.setWidth(rectangle.width + 1);
}

void main(List<String> args) {
  var rectangle1 = Rectangle(10, 2);
  var square = Square(5, 5);

  increaseReactangleWidth(rectangle1);
  increaseReactangleWidth(square);

  print(rectangle1.area());
  print(square.area());
}

File: 03_liskov_substitution/typescript/bad/index01.bad.ts
class Rectangle {
  constructor(width: number, height: number) {
    this.width = width
    this.height = height
  }
  width: number
  height: number

  setWidth (width: number) {
    this.width = width
  }
  setHeight (height: number) {
    this.height = height
  }

  area (): number {
    return this.width * this.height
  }
}
class Square extends Rectangle {
  setWidth (width: number) {
    this.width = width
    this.height = width
  }
  setHeight (height: number) {
    this.width = height
    this.height = height
  }
}

function increaseReactangleWidth (rectangle: Rectangle) {
  rectangle.setWidth(rectangle.width + 1)
}

const rectangle1 = new Rectangle(10, 2)
const square = new Square(5, 5)

increaseReactangleWidth(rectangle1)
increaseReactangleWidth(square)

console.log(rectangle1.area())
console.log(square.area())
